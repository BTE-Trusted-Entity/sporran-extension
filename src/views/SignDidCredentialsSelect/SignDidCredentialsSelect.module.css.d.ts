export const __esModule: true;
export const container: string;
export const heading: string;
export const subline: string;
export const credentials: string;
export const identityLine: string;
export const list: string;
export const buttonsLine: string;
export const cancel: string;
export const next: string;
