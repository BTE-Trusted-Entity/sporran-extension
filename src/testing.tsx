export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
