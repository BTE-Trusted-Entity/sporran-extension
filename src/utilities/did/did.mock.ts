import { isFullDid } from './did';

jest.mock('./did');

export function mockIsFullDid(boolean: boolean): void {
  jest.mocked(isFullDid).mockReturnValue(boolean);
}
