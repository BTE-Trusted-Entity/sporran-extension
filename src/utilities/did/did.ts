import {
  ConfigService,
  Did,
  DidDocument,
  DidEncryptionKey,
  DidUri,
  Utils,
} from '@kiltprotocol/sdk-js';

import { useAsyncValue } from '../useAsyncValue/useAsyncValue';

export function isFullDid(did: DidUri | undefined): boolean {
  if (!did) {
    // could be a legacy identity without DID
    return false;
  }
  return Did.parse(did).type === 'full';
}

export async function getFullDidDocument(did: DidUri): Promise<DidDocument> {
  const api = ConfigService.get('api');
  return Did.linkedInfoFromChain(await api.call.did.query(Did.toChain(did)))
    .document;
}

export function useFullDidDocument(did: DidUri): DidDocument | undefined {
  return useAsyncValue(getFullDidDocument, [did]);
}

export async function getDidDocument(did: DidUri): Promise<DidDocument> {
  return isFullDid(did)
    ? await getFullDidDocument(did)
    : Did.parseDocumentFromLightDid(did);
}

export function getDidEncryptionKey({
  keyAgreement,
}: DidDocument): DidEncryptionKey {
  if (!keyAgreement || !keyAgreement[0]) {
    throw new Error('encryptionKey is not defined somehow');
  }
  return keyAgreement[0];
}

export function parseDidUri(did: DidUri): ReturnType<typeof Did.parse> & {
  fullDid: DidUri;
} {
  const parsed = Did.parse(did);
  const fullDid = parsed.type === 'full' ? did : Did.getFullDidUri(did);

  return {
    ...parsed,
    fullDid,
  };
}

export async function needLegacyDidCrypto(
  did: DidUri | undefined,
): Promise<boolean> {
  if (!did) {
    // DID was deactivated, no action needed.
    return false;
  }
  if (!isFullDid(did)) {
    return false;
  }

  try {
    const encryptionKey = getDidEncryptionKey(await getDidDocument(did));
    return (
      Utils.Crypto.u8aToHex(encryptionKey.publicKey) ===
      '0xf2c90875e0630bd1700412341e5e9339a57d2fefdbba08de1cac8db5b4145f6e'
    );
  } catch {
    // getDidDocument might throw if the DID is not on-chain anymore (removed, another endpoint),
    // no legacy crypto needed in that case
    return false;
  }
}
