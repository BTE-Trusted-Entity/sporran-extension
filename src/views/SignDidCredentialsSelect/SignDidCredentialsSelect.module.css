.container {
  composes: container from '../../components/View/View.module.css';
}

.heading {
  composes: screenHeadline from '../../components/Typography/Typography.module.css';
}

.subline {
  composes: subline from '../../components/Typography/Typography.module.css';
  margin: 0;
}

.credentials {
  composes: fullWidthContainer from '../../components/View/View.module.css';
  height: calc(370rem / 16);
  padding: 0;
  margin-top: 0.75rem;
}

.identityLine {
  composes: component from '../../components/IdentityLine/IdentityLine.module.css';
  padding-inline: 1.5rem;
}

.list {
  list-style: none;
  padding: 0;
  margin: 0;
}

.buttonsLine {
  composes: buttonsLine from '../../components/Button/Button.module.css';
}

.cancel {
  composes: buttonSecondary from '../../components/Button/Button.module.css';
}

.next {
  composes: buttonPrimary from '../../components/Button/Button.module.css';
}
