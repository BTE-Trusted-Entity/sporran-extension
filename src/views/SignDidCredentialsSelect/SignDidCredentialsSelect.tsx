import { browser } from 'webextension-polyfill-ts';

import { useRef, useState, useCallback } from 'react';

import { omit } from 'lodash-es';

import * as styles from './SignDidCredentialsSelect.module.css';

import { Identity } from '../../utilities/identities/types';
import {
  SharedCredential,
  useIdentityCredentials,
} from '../../utilities/credentials/credentials';

import { IdentityLine } from '../../components/IdentityLine/IdentityLine';
import { SignDidCredentialCard } from '../../components/CredentialCard/SignDidCredentialCard';
import { LinkBack } from '../../components/LinkBack/LinkBack';

interface Props {
  identity: Identity;
  onCancel: () => void;
  onSubmit: (selected: SharedCredential[]) => void;
}

export function SignDidCredentialsSelect({
  identity,
  onCancel,
  onSubmit,
}: Props): JSX.Element | null {
  const t = browser.i18n.getMessage;

  const sporranCredentials = useIdentityCredentials(identity.did);

  const ref = useRef<HTMLElement>(null);

  const [selected, setSelected] = useState<Record<string, SharedCredential>>();

  const onSelect = useCallback(
    (presentation: SharedCredential) => {
      const key = presentation.sporranCredential.credential.rootHash;
      setSelected({ ...selected, [key]: presentation });
    },
    [selected],
  );

  const onUnSelect = useCallback(
    (key: string) => {
      setSelected(omit(selected, key));
    },
    [selected],
  );

  const handleSubmit = useCallback(() => {
    if (!selected) {
      return;
    }
    onSubmit(Object.values(selected));
  }, [selected, onSubmit]);

  return (
    <form className={styles.container} onSubmit={handleSubmit}>
      <h1 className={styles.heading}>
        {t('view_SignDidCredentialsSelect_heading')}
      </h1>
      <p className={styles.subline}>
        {t('view_SignDidCredentialsSelect_subline')}
      </p>

      <section className={styles.credentials} ref={ref}>
        <IdentityLine identity={identity} className={styles.identityLine} />

        <ul className={styles.list}>
          {sporranCredentials?.map((sporranCredential) => (
            <SignDidCredentialCard
              key={sporranCredential.credential.rootHash}
              sporranCredential={sporranCredential}
              onSelect={onSelect}
              onUnSelect={onUnSelect}
              viewRef={ref}
            />
          ))}
        </ul>
      </section>

      <p className={styles.buttonsLine}>
        <button type="button" className={styles.cancel} onClick={onCancel}>
          {t('common_action_cancel')}
        </button>
        <button
          type="submit"
          className={styles.next}
          disabled={!selected || !Object.entries(selected).length}
        >
          {t('view_SignDidCredentialsSelect_next')}
        </button>
      </p>

      <LinkBack />
    </form>
  );
}
