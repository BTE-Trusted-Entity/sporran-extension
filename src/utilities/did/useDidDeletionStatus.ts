import { IDidDetails } from '@kiltprotocol/types';
import { DidResolver } from '@kiltprotocol/did';

import { useAsyncValue } from '../useAsyncValue/useAsyncValue';

async function getDidDeletionStatus(did: IDidDetails['did']): Promise<boolean> {
  if (!did) {
    return false;
  }

  try {
    const resolved = await DidResolver.resolveDoc(did);
    return Boolean(
      resolved && resolved.metadata && resolved.metadata.deactivated,
    );
  } catch (error) {
    console.error(error);
    throw new Error('Could not get DID deletion status');
  }
}

export function useDidDeletionStatus(
  did: IDidDetails['did'],
): boolean | undefined {
  return useAsyncValue(getDidDeletionStatus, [did]);
}
