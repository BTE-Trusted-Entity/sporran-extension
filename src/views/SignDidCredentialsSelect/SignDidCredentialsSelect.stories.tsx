import { Meta } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import { identitiesMock as identities } from '../../utilities/identities/IdentitiesProvider.mock';
import { mockRequestCredential } from '../../utilities/credentials/CredentialsProvider.mock';
import { PopupTestProvider } from '../../utilities/popups/PopupTestProvider';

import { paths } from '../paths';

import { SignDidCredentialsSelect } from './SignDidCredentialsSelect';

export default {
  title: 'Views/SignDidCredentialsSelect',
  component: SignDidCredentialsSelect,
} as Meta;

export function Template(): JSX.Element {
  return (
    <PopupTestProvider
      path={paths.popup.share.start}
      data={mockRequestCredential}
    >
      <SignDidCredentialsSelect
        identity={
          identities['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo']
        }
        onCancel={action('onCancel')}
        onSubmit={action('onSubmit')}
      />
    </PopupTestProvider>
  );
}
