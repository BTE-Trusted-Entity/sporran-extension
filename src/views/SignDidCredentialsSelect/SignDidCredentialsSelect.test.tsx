import { identitiesMock as identities, render } from '../../testing/testing';

import { parseDidUri } from '../../utilities/did/did';

import { mockIsFullDid } from '../../utilities/did/did.mock';

import { SignDidCredentialsSelect } from './SignDidCredentialsSelect';

jest.mock('../../utilities/did/did');

jest.mocked(parseDidUri).mockReturnValue({
  fullDid: 'did:kilt:4oeJ76hdj84xnwCNqijUHUCTmfwXgSZ4vmxLEiTEYgQdBCcZ',
} as unknown as ReturnType<typeof parseDidUri>);
mockIsFullDid(true);

describe('SignDidCredentialsSelect', () => {
  it('should render', async () => {
    const { container } = render(
      <SignDidCredentialsSelect
        identity={
          identities['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo']
        }
        onCancel={jest.fn()}
        onSubmit={jest.fn()}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
