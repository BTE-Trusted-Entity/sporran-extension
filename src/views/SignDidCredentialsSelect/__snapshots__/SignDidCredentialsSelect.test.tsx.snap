// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignDidCredentialsSelect should render 1`] = `
<div>
  <form
    class="container"
  >
    <h1
      class="heading"
    >
      Select Credentials
    </h1>
    <p
      class="subline"
    >
      Select credentials to attach to your signature
    </p>
    <section
      class="credentials"
    >
      <section
        class="identityLine"
      >
        <section
          class="identity"
        >
          <div
            class="avatar fullDid"
          >
            Identicon
          </div>
          <p
            class="name"
          >
            Identity with on-chain DID and lots of coins
          </p>
        </section>
        <span
          class="amount"
        >
          1.2340
           
          <span
            aria-label="KILT Coin"
            class="coin"
          />
        </span>
      </section>
      <ul
        class="list"
      >
        <li
          aria-expanded="false"
          class="selectable"
        >
          <input
            aria-labelledby="collapsedLabel"
            class="selectMultiple"
            id="0xbe7fcc7aa0186b05fd116c100e2d673fb951163693788640ce6032ad2f700dae"
            name="credential"
            type="checkbox"
          />
          <button
            class="expand"
          >
            <section
              class="collapsedShareCredential"
            >
              <h4
                class="collapsedName"
                id="collapsedLabel"
              >
                Email Credential
              </h4>
              <p
                class="collapsedValue"
              >
                Trusted Entity attester
              </p>
            </section>
          </button>
        </li>
        <li
          aria-expanded="false"
          class="selectable"
        >
          <input
            aria-labelledby="collapsedLabel"
            class="selectMultiple"
            disabled=""
            id="0xbe7fcc7aa0186b05fd116c100e2d673fb951163693788640ce6032ad2f700daf"
            name="credential"
            type="checkbox"
          />
          <button
            class="expand"
          >
            <section
              class="collapsedShareCredential notUsable"
            >
              <h4
                class="collapsedName"
                id="collapsedLabel"
              >
                Alter Ego Email
              </h4>
              <p
                class="collapsedValue"
              >
                Anonymous
              </p>
            </section>
          </button>
        </li>
        <li
          aria-expanded="false"
          class="selectable"
        >
          <input
            aria-labelledby="collapsedLabel"
            class="selectMultiple"
            id="0xbe7fcc7aa0186b05fd116c100e2d673fb951163693788640ce6032ad2f700foo"
            name="credential"
            type="checkbox"
          />
          <button
            class="expand"
          >
            <section
              class="collapsedShareCredential"
            >
              <h4
                class="collapsedName"
                id="collapsedLabel"
              >
                A really long name for my credential which probably will not fit on the screen
              </h4>
              <p
                class="collapsedValue"
              >
                Lorem Ipsum
              </p>
            </section>
          </button>
        </li>
        <li
          aria-expanded="false"
          class="selectable"
        >
          <input
            aria-labelledby="collapsedLabel"
            class="selectMultiple"
            id="0xbe7fcc7aa0186b05fd116c100e2d673fb951163693788640ce6032ad2f700bar"
            name="credential"
            type="checkbox"
          />
          <button
            class="expand"
          >
            <section
              class="collapsedShareCredential"
            >
              <h4
                class="collapsedName"
                id="collapsedLabel"
              >
                Mein Credential
              </h4>
              <p
                class="collapsedValue"
              >
                Beamter #45065-2
              </p>
            </section>
          </button>
        </li>
        <li
          aria-expanded="false"
          class="selectable"
        >
          <input
            aria-labelledby="collapsedLabel"
            class="selectMultiple"
            id="0xbe7fcc7aa0186b05fd116c100e2d673fb951163693788640ce6032ad2f700dah"
            name="credential"
            type="checkbox"
          />
          <button
            class="expand"
          >
            <section
              class="collapsedShareCredential"
            >
              <h4
                class="collapsedName"
                id="collapsedLabel"
              >
                Tweet Twit Twat
              </h4>
              <p
                class="collapsedValue"
              >
                Social KYC
              </p>
            </section>
          </button>
        </li>
        <li
          aria-expanded="false"
          class="selectable"
        >
          <input
            aria-labelledby="collapsedLabel"
            class="selectMultiple"
            id="0xbe7fcc7aa0186b05fd116c100e2d673fb951163693788640ce6032ad2f700daj"
            name="credential"
            type="checkbox"
          />
          <button
            class="expand"
          >
            <section
              class="collapsedShareCredential"
            >
              <h4
                class="collapsedName"
                id="collapsedLabel"
              >
                Dotsama Credential
              </h4>
              <p
                class="collapsedValue"
              >
                Myself
              </p>
            </section>
          </button>
        </li>
        <li
          aria-expanded="false"
          class="selectable"
        >
          <input
            aria-labelledby="collapsedLabel"
            class="selectMultiple"
            id="0xbe7fcc7aa0186b05fd116c100e2d673fb951163693788640ce6032ad2f700dak"
            name="credential"
            type="checkbox"
          />
          <button
            class="expand"
          >
            <section
              class="collapsedShareCredential"
            >
              <h4
                class="collapsedName"
                id="collapsedLabel"
              >
                All your stakes are belong to us
              </h4>
              <p
                class="collapsedValue"
              >
                Foo
              </p>
            </section>
          </button>
        </li>
        <li
          aria-expanded="false"
          class="selectable"
        >
          <input
            aria-labelledby="collapsedLabel"
            class="selectMultiple"
            id="0xbe7fcc7aa0186b05fd116c100e2d673fb951163693788640ce6032ad2f700dal"
            name="credential"
            type="checkbox"
          />
          <button
            class="expand"
          >
            <section
              class="collapsedShareCredential"
            >
              <h4
                class="collapsedName"
                id="collapsedLabel"
              >
                Lots of props
              </h4>
              <p
                class="collapsedValue"
              >
                Trustity entity attester
              </p>
            </section>
          </button>
        </li>
        <li
          aria-expanded="false"
          class="selectable"
        >
          <input
            aria-labelledby="collapsedLabel"
            class="selectMultiple"
            id="0xbe7fcc7aa0186b05fd116c100e2d673fb951163693788640ce6032ad2f700dan"
            name="credential"
            type="checkbox"
          />
          <button
            class="expand"
          >
            <section
              class="collapsedShareCredential"
            >
              <h4
                class="collapsedName"
                id="collapsedLabel"
              >
                Email Credential
              </h4>
              <p
                class="collapsedValue"
              >
                Trusted Entity attester
              </p>
            </section>
          </button>
        </li>
        <li
          aria-expanded="false"
          class="selectable"
        >
          <input
            aria-labelledby="collapsedLabel"
            class="selectMultiple"
            id="0xe4cc9f4eebf1112b3f26ad6bde36ef95ea79d544c481270f62a50d699fac819b"
            name="credential"
            type="checkbox"
          />
          <button
            class="expand"
          >
            <section
              class="collapsedShareCredential"
            >
              <h4
                class="collapsedName"
                id="collapsedLabel"
              >
                GitHub 1
              </h4>
              <p
                class="collapsedValue"
              >
                SocialKYC
              </p>
            </section>
          </button>
        </li>
        <li
          aria-expanded="false"
          class="selectable"
        >
          <input
            aria-labelledby="collapsedLabel"
            class="selectMultiple"
            id="0x86a20051e6d1eecb0dd6e9c7b3b0cf946867e662b525ef2e9afe8428f6616ae0"
            name="credential"
            type="checkbox"
          />
          <button
            class="expand"
          >
            <section
              class="collapsedShareCredential"
            >
              <h4
                class="collapsedName"
                id="collapsedLabel"
              >
                Not downloaded Email Credential
              </h4>
              <p
                class="collapsedValue"
              >
                Trusted Entity attester
              </p>
            </section>
          </button>
        </li>
      </ul>
    </section>
    <p
      class="buttonsLine"
    >
      <button
        class="cancel"
        type="button"
      >
        Cancel
      </button>
      <button
        class="next"
        disabled=""
        type="submit"
      >
        Next
      </button>
    </p>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
  </form>
</div>
`;
